libsweep/sweep.cc

```diff
@@ -42,10 +22,34 @@ const char* sweep_error_message(sweep_error_s error) {
 
 void sweep_error_destruct(sweep_error_s error) {
   SWEEP_ASSERT(error);
-
   delete error;
 }
 
+template <class F> auto translateException(F&& f, sweep_error_s* error) -> decltype(std::forward<F>(f)()) {
+  try {
+    return std::forward<F>(f)();
+  } catch (const sweep::ErrorBase& e) {
+    *error = sweep_error_construct(e.what());
+  } catch (const std::bad_alloc&) {
+    *error = sweep_error_construct("Could not allocate enough memory");
+  } catch (...) {
+    *error = sweep_error_construct("Unknown exception");
+  }
+  return {};
+}
+
+template <class F> void translateExceptionV(F&& f, sweep_error_s* error) {
```

Comment:
Yes, that was my first intention but I couldn't think of a simple SFINAE expression from the top of my head, so I went with the *manual* solution